rfcs/2021-05-23-groups2.md

```diff
@@ -0,0 +1,134 @@
+- message `G:gId MEM [C:cId]` - connection created with all group members for a given member or current client
+- command `G:gId SEND msg` - send message to group
+- message `G:gId SENT [C:cId] msgId` - notification that the message is sent and its internal ID
+- message `G:gId MSG C:cId msgdata` - received group message from cId, msgdata is the same set of parameters as in `MSG`
+- command `G:gId ACK msgId` - acknowledge message reception by the client
+- message `G:gId RCVD [C:cId] msgId status` - message delivery notification
+- command `G:gId LEAVE` - leave the group
+- message `G:gId LEFT [C:cId]` - connection cId left the group
+- command `G:gId REM C:cId` - remove group member (response is `G:gId OK`, followed by `REMD` notification)
+- message `G:gId REMD C:cId [C:cId]` - member removed (who, by whom - if it's not the current user)
+- message `G:gId OUT C:cId` - you are removed
+- message `G:gId EMPTY` - all members left the group and it is now empty
+- command `G:gId DEL` - delete the group (response is `G:gId OK`)
+- message `G:gId DELD [C:cId]` - group deleted (by whom, if it's not the current user)
+- command `G:bId LS` - list connections in broadcast, response is `MEM space_separated_connections`
+- message `G:bId MS space_separated_connections`
+
+## Agent message envelopes syntax (group-specific)
+
+- `GROUP C:mid G:inv gInfo` - invitation to join the group
+- `MEM C:mid` - confirmation that member connected to all members
+- `LEFT` - notification that member left the group
+- `OUT` - you are removed from the group
+- `REM C:mid` - remove member mid from the group
+- `REMD C:mid` - confirmation that member is removed
+- `DEL` - group is deleted
+- `DELD` - confirmation that group is deleted
+
+## Protocol costs
+
+- Adding a member:
+  - GROUP - 1
+  - introductions - (4 + connection costs) * (n - 1)
+  - MEM - (n - 1)
+  - total - `6n-4` messages
+
+- Sending a message:
+  - MSG - (n-1)
+  - RCVD - (n-1)
+  - total - `2n-2` messages
+
+- Member leaves a group:
+  - LEFT - (n-1)
+  - total -`n-1` messages
+
+- Member is removed by another member:
+  - OUT - 1
+  - REM - (n-1)
+  - REMD - (n-1)
+  - total `2n-1` messages
+
+- Group is deleted:
+  - DEL - (n-1)
+  - DELD - (n-1)
+  - total - `2n-2` messages
+
+- Group integrity verification:
+  - TODO
+
+## Questions
+
+1. Message verification keys. Agents use separate server and encryption key for each connection, but there can be a value of having the same message verification key used for all members in the group. E.g., a member can validate to other group members that the message was delivered to all group members by sending signed message receipts they receive from the agent. This would mean that for each message `n*(n-1)` messages will be send, although signed receipts can be grouped to reduce this number. It can be done periodically, rather than on each message, as described [here](https://signal.org/blog/private-groups/).
```

Comment:
`member can validate to other group members that the message was delivered to all group members by sending signed message receipts they receive from the agent`
do you mean e.g.
1) m1 sends msg to m2, m3
2) m2 sends signed receipt back to m1
3) m3 sends signed receipt back to m1
4) m1 sends both signed receipts to m2, m3
It doesn't look like message integrity schemes described before, where either m1 regularly sends digests (without messages from m2,m3) or only 2) 3) happens without 4)

Reply:
> It doesn't look like message integrity schemes described before, where either m1 regularly sends digests (without messages from m2,m3) or only 2) 3) happens without 4)

it is, roughly one of two schemes.